examples/stable_diffusion_xl/gm/models/trainer_factory.py

```diff
@@ -153,31 +153,30 @@ def __init__(
         self.clip_grad = clip_grad
         self.clip_norm = clip_norm
 
-        self.enable_accum_grad = False
-        if grad_accum_steps > 1:
-            self.enable_accum_grad = True
-            self.accum_steps = grad_accum_steps
-            self.accum_step = ms.Parameter(ms.Tensor(0, dtype=ms.int32), name="accum_step")
+        self.accum_steps = grad_accum_steps
+        if self.accum_steps > 1:
+            self.accum_step = ms.Parameter(ms.Tensor(0, dtype=ms.int32), name="accum_step", requires_grad=False)
             self.accumulated_grads = optimizer.parameters.clone(prefix="accum_grad", init="zeros")
             self.hyper_map = ops.HyperMap()
 
     def do_optim(self, loss, grads):
-        if not self.enable_accum_grad:
+        if not self.accum_steps > 1:
             if self.clip_grad:
                 grads = ops.clip_by_global_norm(grads, self.clip_norm)
             loss = F.depend(loss, self.optimizer(grads))
         else:
-            self.accum_step += 1
             loss = F.depend(
                 loss, self.hyper_map(F.partial(_grad_accum_op, self.accum_steps), self.accumulated_grads, grads)
             )
+            loss = F.depend(loss, ops.assign_add(self.accum_step, ms.Tensor(1, ms.int32)))
             if self.accum_step % self.accum_steps == 0:
                 if self.clip_grad:
                     grads = ops.clip_by_global_norm(self.accumulated_grads, self.clip_norm)
                     loss = F.depend(loss, self.optimizer(grads))
                 else:
                     loss = F.depend(loss, self.optimizer(self.accumulated_grads))
                 loss = F.depend(loss, self.hyper_map(F.partial(_grad_clear_op), self.accumulated_grads))
+                loss = F.depend(loss, ops.assign(self.accum_step, ms.Tensor(0, ms.int32)))
```

Comment:
Setting `self.accum_step` to `0` will turn off gradient accumulation due to the `if not self.accum_steps > 1` condition. I think this statement can be omitted.

Reply:
No worry. `self.accum_steps` is a constant scalar representing the the number of gradient accumulation steps, e.g., 4. `self.accum_step` is a ms.Tensor that records the current accumulated steps.